module/zfs/dmu_send.c

```diff
@@ -3720,6 +3746,38 @@ dmu_recv_stream(dmu_recv_cookie_t *drc, vnode_t *vp, offset_t *voffp,
 	}
 	mutex_exit(&rwa->mutex);
 
+	/*
+	 * if we are receiving a full stream as clone, we need to free leftover
+	 * objects after the last one referenced in the stream.
+	 */
+	if (drc->drc_clone && drc->drc_drrb->drr_fromguid == 0) {
+		uint64_t obj;
+		int free_err = 0, next_err = 0;
+
+		for (obj = rwa->last_touched_object + 1;
+		    next_err == 0;
+		    next_err = dmu_object_next(rwa->os, &obj, FALSE, 0)) {
```

Comment:
nit: I think rewriting this in the form `while (next_err == 0) { ... }` would be easier to read.

Reply:
that was lifted from receive_freeobjects (https://github.com/zfsonlinux/zfs/blob/master/module/zfs/dmu_send.c#L2550), but since the loop condition is much simpler here converting to a while makes sense.

Reply:
on second thought - with the pesky continue we'd either need two calls to dmu_object_next or a goto, so I am not so sure..

https://gist.github.com/Fabian-Gruenbichler/8632d82204237a6219ba40c1258d261d
https://gist.github.com/Fabian-Gruenbichler/6234f96c6a33397561541f75889a7d3b

which of the three would be preferred?

Reply:
That makes sense, my preference would be for the last one
https://gist.github.com/Fabian-Gruenbichler/6234f96c6a33397561541f75889a7d3b